### ISSUE TITLE
dns over tls with systemd-resolved not work

### ISSUE THREAD

**Issue by dasaferi** [nextdns/nextdns#557]
**Context**
 - CLI Version: systemd 245 (245.4-4ubuntu3.11)
 - Platform: Ubuntu 20.04.2 LTS x86_64
 - Kernel: 5.8.0-63-generic

Hi, I have setup systemd-resolved as from instructions supplied from the site editing the file /etc/systemd/resolved.conf with nano as sudo user. Once edited I restarted the pc to apply the changes, but no noticed that the pc could not resolve the domains, I tried also with ping google.com and it appeared a writing in the console impossible to resolve the domain.
```
[Resolve]
DNS=45.90.28.0#dcc2e7.dns1.nextdns.io
DNS=2a07:a8c0::#dcc2e7.dns1.nextdns.io
DNS=45.90.30.0#dcc2e7.dns2.nextdns.io
DNS=2a07:a8c1::#dcc2e7.dns2.nextdns.io
DNSOverTLS=yes
```
Searching the internet I saw that some users in 2020 solved it by changing #Domains to Domains=~. , I did so but I always had the same problem impossible to solve the domain in the console.

```
[Resolve]
DNS=45.90.28.0#dcc2e7.dns1.nextdns.io
DNS=2a07:a8c0::#dcc2e7.dns1.nextdns.io
DNS=45.90.30.0#dcc2e7.dns2.nextdns.io
DNS=2a07:a8c1::#dcc2e7.dns2.nextdns.io
Domains=~.
DNSOverTLS=yes
```

Reading the documentation of [arch linux ](https://wiki.archlinux.org/title/Systemd-resolved) I saw that you had to set also DNSSEC=yes, so I did, I restarted the pc I made ping to google.com and always impossible to resolve the domain.

```
[Resolve]
DNS=45.90.28.0#dcc2e7.dns1.nextdns.io
DNS=2a07:a8c0::#dcc2e7.dns1.nextdns.io
DNS=45.90.30.0#dcc2e7.dns2.nextdns.io
DNS=2a07:a8c1::#dcc2e7.dns2.nextdns.io
Domains=~.
DNSSEC=yes
DNSOverTLS=yes
```

At this point I tried to put `dcc2e7.dns.nextdns.io` only and I restarted the pc and I pinged google.com and always impossible to resolve the domain, can you kindly try to set on ubuntu nextdns through systemd-resolved? I tried everything but I could not do it.
I also tried systemd-resolved with cloudflare, google and quad9 dns (9.9.9.9#dns.quad9.net) and everything worked, i don't understand why nextdns doesn't work through systemd-resolved.

PS. the dns data i provided in this issue (account name and ip) are temporary since 7 days (they were generated by https://my.nextdns.io/start without being logged in), i didn't write the real user name and ip.

Thanks for your patience and sorry for the inconvenience, have a nice day.

**Comment by vincentbernat** [comment]
Can you try with `resolvectl query` instead of `ping`? It may display more information.

**Comment by dasaferi** [comment]
@vincentbernat hello, thanks for the help. I did in console `resolvectl query` and this is the output.
`google.com: resolve call failed: Lookup failed due to system error: Connection timed out`

I also did the `resolvectl status` command in console and this is the output.
```
Global
       LLMNR setting: no                  
MulticastDNS setting: no                  
  DNSOverTLS setting: yes                 
      DNSSEC setting: yes                 
    DNSSEC supported: yes                 
  Current DNS Server: 45.90.30.0          
         DNS Servers: 45.90.28.0          
                      2a07:a8c0::         
                      45.90.30.0          
                      2a07:a8c1::         
          DNS Domain: ~.                  
          DNSSEC NTA: 10.in-addr.arpa     
                      16.172.in-addr.arpa 
                      168.192.in-addr.arpa
                      17.172.in-addr.arpa 
                      18.172.in-addr.arpa 
                      19.172.in-addr.arpa 
                      20.172.in-addr.arpa 
                      21.172.in-addr.arpa 
                      22.172.in-addr.arpa 
                      23.172.in-addr.arpa 
                      24.172.in-addr.arpa 
                      25.172.in-addr.arpa 
                      26.172.in-addr.arpa 
                      27.172.in-addr.arpa 
                      28.172.in-addr.arpa 
                      29.172.in-addr.arpa 
                      30.172.in-addr.arpa 
                      31.172.in-addr.arpa 
                      corp                
                      d.f.ip6.arpa        
                      home                
                      internal            
                      intranet            
                      lan                 
                      local               
                      private             
                      test
```
I noticed that in the entry dns is truncated the dns I entered is missing the part `#dcc2e7.dns1.nextdns.io` and `#dcc2e7.dns2.nextdns.io` 🤔, at this point I did `cat /etc/systemd/resolved.conf` and everything seems to be written correctly.

```
#  This file is part of systemd.
#
#  systemd is free software; you can redistribute it and/or modify it
#  under the terms of the GNU Lesser General Public License as published by
#  the Free Software Foundation; either version 2.1 of the License, or
#  (at your option) any later version.
#
# Entries in this file show the compile time defaults.
# You can change settings by editing this file.
# Defaults can be restored by simply deleting this file.
#
# See resolved.conf(5) for details

[Resolve]
DNS=45.90.28.0#dcc2e7.dns1.nextdns.io
DNS=2a07:a8c0::#dcc2e7.dns1.nextdns.io
DNS=45.90.30.0#dcc2e7.dns2.nextdns.io
DNS=2a07:a8c1::#dcc2e7.dns2.nextdns.io
#FallbackDNS=
Domains=~.
#LLMNR=no
#MulticastDNS=no
DNSSEC=yes
DNSOverTLS=yes
#Cache=no-negative
#DNSStubListener=yes
#ReadEtcHosts=yes
```
I hope I have given as much information as possible to solve the problem, thanks again.

**Comment by vincentbernat** [comment]
I have tried with your configuration and systemd 246 and it works fine. Notably, `resolvectl status` shows the complete address (with the name for SNI). The ability to configure this way is new with 246:

        * systemd-resolved's DNS= configuration option now optionally accepts a
          port number (after ":") and a host name (after "#"). When the host
          name is specified, the DNS-over-TLS certificate is validated to match
          the specified hostname. Additionally, in case of IPv6 addresses, an
          interface may be specified (after "%").

You can either upgrade to Ubuntu 21.04 (systemd 247) or use nextdns own client and tell systemd-resolved to use it. Make it listen to another port, like 5553 and use:

```ini
[Resolve]
DNS=127.0.0.1:8553
Domains=~.
```

Where did you see the documentation about resolved.conf? We should update it to put the minimum version.

**Comment by dasaferi** [comment]
Hi, thanks for the help given to me and thanks for trying the dns via systemd. I saw the documentation from arch linux https://wiki.archlinux.org/title/Systemd-resolved I also found the documentation of fedora https://fedoraproject.org/wiki/Changes/systemd-resolved only it doesn't contain much information, the documentation on systemd-resolved is really poor.

I will do as you advised through the official client and set systemd to listen for local requests on port 8553, thank you very much.
